From 457ada304915fc656d27685a22c620379eb8d487 Mon Sep 17 00:00:00 2001
From: Alisher Alikhodjaev <alisher@google.com>
Date: Tue, 24 Jan 2023 21:47:07 +0000
Subject: [PATCH] Revert "OOBR in NxpMfcReader::SendIncDecRestoreCmdPart2"

This reverts commit 4d23eb81604ceea12d2929fe1f78873d6d604d1c.

Reason for revert: CVE-2022-20471 re-released due to functional regression

Change-Id: I374969dbca3adb69100b516be3209683fe891729
Merged-In: I188dac7bd3475e96b6c78ef5992b6422522b205e
(cherry picked from commit 117beb07e435ff84c74bf4e8dcb03ba8241e9a62)
Merged-In: I374969dbca3adb69100b516be3209683fe891729
---
 halimpl/mifare/NxpMfcReader.cc | 10 ++--------
 halimpl/mifare/NxpMfcReader.h  |  4 ++--
 2 files changed, 4 insertions(+), 10 deletions(-)

diff --git a/halimpl/mifare/NxpMfcReader.cc b/halimpl/mifare/NxpMfcReader.cc
index 6277f3e..f4a1878 100644
--- a/halimpl/mifare/NxpMfcReader.cc
+++ b/halimpl/mifare/NxpMfcReader.cc
@@ -55,7 +55,7 @@ int NxpMfcReader::Write(uint16_t mfcDataLen, const uint8_t *pMfcData) {
   /* send TAG_CMD part 2 for Mifare increment ,decrement and restore commands */
   if (mfcTagCmdBuff[4] == eMifareDec || mfcTagCmdBuff[4] == eMifareInc ||
       mfcTagCmdBuff[4] == eMifareRestore) {
-    SendIncDecRestoreCmdPart2(mfcDataLen, pMfcData);
+    SendIncDecRestoreCmdPart2(pMfcData);
   }
   return writtenDataLen;
 }
@@ -259,20 +259,14 @@ void NxpMfcReader::AuthForWrite() {
 ** Returns          None
 **
 *******************************************************************************/
-void NxpMfcReader::SendIncDecRestoreCmdPart2(uint16_t mfcDataLen,
-                                             const uint8_t *mfcData) {
+void NxpMfcReader::SendIncDecRestoreCmdPart2(const uint8_t *mfcData) {
   NFCSTATUS status = NFCSTATUS_SUCCESS;
   /* Build TAG_CMD part 2 for Mifare increment ,decrement and restore commands*/
   uint8_t incDecRestorePart2[] = {0x00, 0x00, 0x05, (uint8_t)eMfRawDataXchgHdr,
                                   0x00, 0x00, 0x00, 0x00};
   uint8_t incDecRestorePart2Size =
       (sizeof(incDecRestorePart2) / sizeof(incDecRestorePart2[0]));
-
   if (mfcData[3] == eMifareInc || mfcData[3] == eMifareDec) {
-    if (incDecRestorePart2Size >= mfcDataLen) {
-      incDecRestorePart2Size = mfcDataLen - 1;
-      android_errorWriteLog(0x534e4554, "238177877");
-    }
     for (int i = 4; i < incDecRestorePart2Size; i++) {
       incDecRestorePart2[i] = mfcData[i + 1];
     }
diff --git a/halimpl/mifare/NxpMfcReader.h b/halimpl/mifare/NxpMfcReader.h
index 29d40f2..da216e1 100644
--- a/halimpl/mifare/NxpMfcReader.h
+++ b/halimpl/mifare/NxpMfcReader.h
@@ -109,7 +109,7 @@ private:
   void BuildIncDecCmd();
   void CalcSectorAddress();
   void AuthForWrite();
-  void SendIncDecRestoreCmdPart2(uint16_t mfcDataLen, const uint8_t* mfcData);
+  void SendIncDecRestoreCmdPart2(const uint8_t *mfcData);
 
 public:
   int Write(uint16_t mfcDataLen, const uint8_t *pMfcData);
@@ -117,4 +117,4 @@ public:
   NFCSTATUS CheckMfcResponse(uint8_t *pTransceiveData,
                              uint16_t transceiveDataLen);
   static NxpMfcReader &getInstance();
-};
+};
\ No newline at end of file
-- 
2.39.1.456.gfc5497dd1b-goog

